src/schema/formats.js

```diff
@@ -43,12 +43,52 @@ export function isValidVersionString(version) {
   return true;
 }
 
-export function isRelativeURL(url) {
-  let parsed = parse(url);
+export function isAbsoluteUrl(value) {
+  try {
+    new URL(value);
+    return true;
+  } catch (e) {
+    return false;
+  }
+}
+
+export function isAnyUrl(value) {
+  return isAbsoluteUrl(value) || isRelativeUrl(value);
+}
+
+export function isStrictRelativeUrl(value) {
+  return !value.startsWith('//') && isRelativeUrl(value);
+}
 
-  if (parsed.protocol !== '' || parsed.href.startsWith('//')) {
+export function isSecureUrl(value) {
+  try {
+    const url = new URL(value);
+    return ['https:', 'wss:'].includes(url.protocol);
+  } catch (e) {
     return false;
   }
+}
 
-  return true;
+function getRandomProtocol() {
+  const offset = 'a'.charCodeAt(0);
+  const length = 10;
+  const codes = new Array(length);
+  for (let i = 0; i < length; i++) {
+    codes[i] = Math.floor(Math.random() * 26) + offset;
+  }
+  const protocol = String.fromCharCode(...codes);
+  return `${protocol}:`;
+}
+
+function isRelativeUrl(value) {
+  const protocol = getRandomProtocol();
```

Comment:
Why does this have to be random? Can't we just use something predictable that get's re-used every time?